service/runtime/tunnel.go

```diff
@@ -0,0 +1,262 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package runtime
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/gookit/color"
+	colorjson "github.com/neilotoole/jsoncolor"
+	"github.com/pkg/errors"
+	"github.com/upvestco/httpsignature-proxy/service/logger"
+	"golang.org/x/exp/maps"
+)
+
+type tunnel struct {
+	apiClient    ApiClient
+	eventsFilter map[string]interface{}
+	logger       logger.Logger
+	logHeaders   bool
+	cancel       context.CancelFunc
+}
+
+var cyan = fmt.Sprintf
+var lightRed = fmt.Sprintf
+
+func init() {
+	if colorjson.IsColorTerminal(os.Stdout) {
+		cyan = color.FgCyan.Sprintf
+		lightRed = color.FgLightRed.Sprintf
+	}
+}
+
+func createTunnel(apiClient ApiClient, events []string, logHeaders bool, logger logger.Logger) *tunnel {
+	eventsFilter := map[string]interface{}{}
+	for _, t := range events {
+		if len(t) == 0 {
+			continue
+		}
+		eventsFilter[t] = 1
+	}
+
+	return &tunnel{
+		apiClient:    apiClient,
+		eventsFilter: eventsFilter,
+		logger:       logger,
+		logHeaders:   logHeaders,
+	}
+}
+
+const requiredScopes = "webhooks:admin"
+
+func (e *tunnel) doPulling(ctx context.Context, endpointID string) error {
+
+	e.logger.LogF("start pulling evens")
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			if err := e.pullEvents(ctx, endpointID); err != nil {
+				if errors.Is(err, context.Canceled) {
+					return nil
+				}
+				if errors.Is(err, syscall.ECONNREFUSED) {
+					return nil
+				}
+				return errors.Wrap(err, "pullEvents")
+			}
+		}
+	}
+}
+
+func (e *tunnel) pullEvents(ctx context.Context, endpointID string) error {
+	items, code, err := e.apiClient.GetEvents(ctx, endpointID)
+	if err != nil {
+		return errors.Wrap(err, "doPull")
+	}
+	if code == http.StatusUnauthorized {
+		if err := e.apiClient.Authorise(ctx, requiredScopes); err != nil {
+			return errors.Wrap(err, "Could not open the Webhook events tunnel. You client must have '"+requiredScopes+"' scope(s)")
+		}
+		items, code, err = e.apiClient.GetEvents(ctx, endpointID)
+		if err != nil {
+			return errors.Wrap(err, "doPull")
+		}
+	}
+	if code != http.StatusOK {
+		return errors.New("unexpected http response code: " + strconv.Itoa(code))
+	}
+
+	for _, item := range items {
+		formatted, origLen, filteredLen := e.filterAndFormat(item.Payload)
+		if filteredLen == 0 {
+			continue
+		}
+		filtered := origLen != filteredLen
+
+		e.logger.PrintLn(cyan("== new incoming webhook request"))
+		e.logger.PrintLn(cyan("== received at: %s", item.CreatedAt.Format(time.DateTime)))
```

Comment:
lets change to: "== new webhook event received == "

Reply:
Done